xds/internal/balancer/priority/balancer_priority.go

```diff
@@ -0,0 +1,360 @@
+ * limitations under the License.
+ */
+
+package priority
+
+import (
+	"errors"
+	"time"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+)
+
+var (
+	errAllPrioritiesRemoved    = errors.New("no locality is provided, all priorities are removed")
+	defaultPriorityInitTimeout = 10 * time.Second
+)
+
+// syncPriority handles priority after a config update. It makes sure the
+// balancer state (started or not) is in sync with the priorities (even in
+// tricky cases where a child is moved from a priority to another).
+//
+// It's guaranteed that after this function returns:
+// - If some child is READY, it is childInUse, and all lower priorities are
+// closed.
+// - If some child is newly started(in Connecting for the first time), it is
+// childInUse, and all lower priorities are closed.
+// - Otherwise, the lowest priority is childInUse (none of the children is
+// ready, and the overall state is not ready).
+//
+// Steps:
+// - If all priorities were deleted, unset childInUse (to an empty string), and
+// set parent ClientConn to TransientFailure
+// - Otherwise, Scan all children from p0, and check balancer stats:
+//   - For any of the following cases:
+// 	   - If balancer is not started (not built), this is either a new child
+//       with high priority, or a new builder for an existing child.
+// 	   - If balancer is READY
+// 	   - If this is the lowest priority
+//   - do the following:
+//     - if this is not the old childInUse, override picker so old picker is no
+//       longer used.
+//     - switch to it (because all higher priorities are neither new or Ready)
+//     - forward the new addresses and config
+//
+// Caller must hold pb.mu.
+func (pb *priorityBalancer) syncPriority() {
+	// Everything was removed by the update.
+	if len(pb.priorities) == 0 {
+		pb.childInUse = ""
+		pb.priorityInUse = 0
+		// Stop the init timer. This can happen if the only priority is removed
+		// shortly after it's added.
+		if timer := pb.priorityInitTimer; timer != nil {
+			timer.Stop()
+			pb.priorityInitTimer = nil
+		}
+		pb.cc.UpdateState(balancer.State{
+			ConnectivityState: connectivity.TransientFailure,
+			Picker:            base.NewErrPicker(errAllPrioritiesRemoved),
+		})
+		return
+	}
+
+	for p, name := range pb.priorities {
+		child, ok := pb.children[name]
+		if !ok {
+			pb.logger.Warningf("child with name %q is not found in children", name)
+			continue
+		}
+
+		if !child.started ||
+			child.state.ConnectivityState == connectivity.Ready ||
+			p == len(pb.priorities)-1 {
+			if pb.childInUse != "" && pb.childInUse != child.name {
+				// childInUse was set and is different from this child, will
+				// change childInUse later. We need to update picker here
+				// immediately so parent stops using the old picker.
+				pb.cc.UpdateState(child.state)
+			}
+			pb.logger.Warningf("switching to (%q, %v) in syncPriority", child.name, p)
```

Comment:
Why is this a warning?  This might need more comments if it's a bad thing, because it isn't obvious that it's bad to me.